File: src/lib/embeddings/local-provider.js
import { pipeline } from '@xenova/transformers';
import { EmbeddingProvider } from './base.js';

export class LocalEmbeddingProvider extends EmbeddingProvider {
  constructor() {
    super();
    this.modelName = process.env.EMBEDDING_MODEL || 'Xenova/all-MiniLM-L6-v2';
    this.dimension = parseInt(process.env.EMBEDDING_DIM || '384', 10);
    this.pipeline = null;
  }

  async ensureLoaded() {
    if (!this.pipeline) {
      console.log(`Initializing local model: ${this.modelName}...`);
      this.pipeline = await pipeline('feature-extraction', this.modelName);
      console.log('Model loaded successfully');
    }
  }

  async generateEmbedding(text) {
    await this.ensureLoaded();
    const output = await this.pipeline(text, {
      pooling: 'mean',
      normalize: true,
    });
    return Array.from(output.data);
  }

  async generateEmbeddings(texts) {
    await this.ensureLoaded();
    const embeddings = await Promise.all(
      texts.map(async (text) => {
        const output = await this.pipeline(text, {
          pooling: 'mean',
          normalize: true,
        });
        return Array.from(output.data);
      })
    );
    return embeddings;
  }

  getDimension() {
    return this.dimension;
  }

  getModelName() {
    return this.modelName;
  }
}

File: scripts-milvus/embeddings/local_provider.py
"""
Local sentence-transformers embedding provider
"""
from .base import EmbeddingProvider
from sentence_transformers import SentenceTransformer
import os
from typing import List


class LocalEmbeddingProvider(EmbeddingProvider):
    def __init__(self):
        self.model_name = os.getenv('EMBEDDING_MODEL', 'all-MiniLM-L6-v2')
        self.model = SentenceTransformer(self.model_name)
        self.dimension = int(os.getenv('EMBEDDING_DIM', '384'))

        print(f"Initialized local provider: {self.model_name} ({self.dimension}D)")

    def get_embedding(self, text: str) -> List[float]:
        return self.model.encode(text).tolist()

    def get_embeddings(self, texts: List[str]) -> List[List[float]]:
        return self.model.encode(texts, show_progress_bar=True).tolist()

    def get_dimension(self) -> int:
        return self.dimension

    def get_model_name(self) -> str:
        return self.model_name
